### PR TITLE
`ref` fields and `scoped`. Modify pages on struct creation, variable declaration, and ref struct.

### LINKED ISSUE #30226: [17.3] `ref` fields and `scoped`
This feature adds `ref` fields and `scoped` declarations to `struct` types.

See the [feature spec](https://github.com/dotnet/csharplang/blob/main/proposals/low-level-struct-improvements.md#provide-ref-fields-and-scoped).

Only the section on *provide ref fields and scoped* have been merged at this time. Other sections are not likely for C# 11.

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #30226 

https://github.com/dotnet/docs/issues/30226#issuecomment-1230840028 describes the tasks in this PR. The commits follow the task list.

**Internal review site:**

- [Language reference article on structs](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/struct?branch=pr-en-us-30912):  `ref struct` content removed.
- [Language reference article on ref struct](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/ref-struct?branch=pr-en-us-30912): Contains `ref struct` content from struct article.
- [Language reference article on `ref` keyword](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/ref?branch=pr-en-us-30912): Update and edit article to add new ref fields.
- [Language reference article on method parameters](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/method-parameters?branch=pr-en-us-30912): New and updated information on pass by value / reference, reference type, value type, and parameter modifiers. Also, define *safe-to-escape* and *ref-safe-to-escape*
- [Language reference article on declaration statements](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/statements/declarations?branch=pr-en-us-30912): New article on variables declarations. Includes `var`, `ref` locals and `ref` fields.
- [Language reference article on assignment operator](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/assignment-operator?branch=pr-en-us-30912): Clarify `ref` assignment.
- [Language reference article on keywords](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords?branch=pr-en-us-30912): add `scoped` contextual keyword

Changes in other files are primarily updating links and link text.

Note for reviewers:  The articles tagged above have the significant changes. Samples moved to the updated folders. Almost all other article changes were updating links to updated content for the language reference.

### CODE REVIEW

File: docs/csharp/language-reference/builtin-types/struct.md [hunk 1]
```diff
@@ -16,6 +16,8 @@ A *structure type* (or *struct type*) is a [value type](value-types.md) that can
 
 [!code-csharp[struct example](snippets/shared/StructType.cs#StructExample)]
 
+`ref struct` types and `readonly ref struct` types are covered in the article on [`ref struct` types](ref-struct.md).
```
Comment (IEvangelist) [inline docs/csharp/language-reference/builtin-types/struct.md:19]:
```suggestion
`ref struct` types and `readonly ref struct` types are covered in the article on [ref struct types](ref-struct.md).
```

We cannot use code fonts in links.

**Comment by BillWagner** [comment]
@Youssef1313 Is there a tag for the F1 help for `scoped`? I'll add it to this PR.

File: docs/csharp/language-reference/keywords/method-parameters.md [hunk 2]
```diff
@@ -10,20 +10,101 @@ ms.assetid: 680e39ff-775b-48b0-9f47-4186a5bfc4a1
...
-Parameters declared for a method without [in](./in-parameter-modifier.md), [ref](./ref.md) or [out](./out-parameter-modifier.md), are passed to the called method by value. That value can be changed in the method, but the changed value will not be retained when control passes back to the calling procedure. By using a method parameter keyword, you can change this behavior.  
-  
- This section describes the keywords you can use when declaring method parameters:  
-  
+In C#, arguments can be passed to parameters either by value or by reference. Remember that C# types can be either reference types (`class`) or value types (`struct`):
+
+- *Pass by value* means **passing a copy of the variable** to the method.
+- *Pass by reference* means **passing access to the variable** to the method.
+- A variable of a *reference type* contains a reference to its data.
+- A variable of a *structure type* contains its data directly.
```
Comment (IEvangelist) [inline docs/csharp/language-reference/keywords/method-parameters.md:18]:
```suggestion
- A variable of a *value type* contains its data directly.
```

I think this should be "value type", I've never heard of "structure type" before this PR. I always thought it was just value or reference types.

File: docs/csharp/language-reference/keywords/method-parameters.md [hunk 3]
```diff
@@ -10,20 +10,101 @@ ms.assetid: 680e39ff-775b-48b0-9f47-4186a5bfc4a1
...
+
+The following example demonstrates passing value-type parameters by value. The variable `n` is passed by value to the method `SquareIt`. Any changes that take place inside the method have no effect on the original value of the variable.
+
+:::code language="csharp" source="./snippets/ParameterModifiers.cs" id="SnippetPassValueByValue":::
+
+The variable `n` is a value type. It contains its data, the value `5`. When `SquareIt` is invoked, the contents of `n` are copied into the parameter `x`, which is squared inside the method. In `Main`, however, the value of `n` is the same after calling the `SquareIt` method as it was before. The change that takes place inside the method only affects the local variable `x`.
+
+## Pass a value type by reference
+
+When you pass a *value* type *by by reference*:
```
Comment (IEvangelist) [inline docs/csharp/language-reference/keywords/method-parameters.md:45]:
```suggestion
When you pass a *value* type *by reference*:
```

File: docs/csharp/language-reference/keywords/ref.md [hunk 4]
```diff
@@ -120,25 +121,41 @@ You can access a value by reference in the same way. In some cases, accessing a
...
+
+In [`ref struct`](../builtin-types/ref-struct.md) types, you can declare fields that are `ref` fields. `ref` fields are valid only in `ref struct` types to ensure the reference doesn't outlive the object it refers to. This feature enables types like <xref:System.Span%601?displayProperty=fullName>:
+
+```csharp
+public readonly ref struct Span<T>
+{
+    internal readonly ref T _reference;
+    private readonly int _length;
+
+    // etc
```
Comment (IEvangelist) [inline docs/csharp/language-reference/keywords/ref.md:154]:
```suggestion
    // Omitted for brevity...
```

File: docs/csharp/language-reference/keywords/snippets/ParameterModifiers.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace keywords;
```
Comment (IEvangelist) [inline docs/csharp/language-reference/keywords/snippets/ParameterModifiers.cs:7]:
```suggestion
namespace Keywords;

```

File: docs/csharp/language-reference/keywords/snippets/PassParameters.cs [hunk 6]
```diff
@@ -1,4 +1,9 @@
-﻿class TheClass
+﻿using System;
+
+namespace keywords;
```
Comment (IEvangelist) [inline docs/csharp/language-reference/keywords/snippets/PassParameters.cs:3]:
```suggestion
namespace Keywords;
```

File: docs/csharp/language-reference/operators/assignment-operator.md [hunk 7]
```diff
@@ -26,15 +26,21 @@ a = (b = c)
 
 The following example demonstrates the usage of the assignment operator with a local variable, a property, and an indexer element as its left-hand operand:
 
-[!code-csharp-interactive[simple assignment](snippets/shared/AssignmentOperator.cs#Simple)]
+:::code language="csharp" source="snippets/shared/AssignmentOperator.cs" id="SnippetSimple":::
 
-## ref assignment operator
+The left operand of an assignment receives the *value* of the right had of the assignment. When the operands are [value types](../builtin-types/value-types.md), assignment copies the contents of the right hand operand. When the operands are [reference types](../builtin-types/reference-types.md), assignment copies the reference to the object.
```
Comment (IEvangelist) [inline docs/csharp/language-reference/operators/assignment-operator.md:31]:
```suggestion
The left operand of an assignment receives the *value* of the right hand of the assignment. When the operands are [value types](../builtin-types/value-types.md), the assignment copies the contents of the right-hand operand. When the operands are [reference types](../builtin-types/reference-types.md), assignment copies the reference to the object.
```

File: docs/csharp/language-reference/statements/declarations.md [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+helpviewer_keywords: 
+  - "var keyword [C#]"
+---
+# Declaration statements
+
+A *declaration statement* declares a new variable, and optionally, initializes it. All variables have declared type. You can learn more about types in the article on the [.NET type system](../../../standard/base-types/common-type-system.md).
+
+## Implicitly typed local variables
+
+Beginning with C# 3, variables that are declared at method scope can have an implicit "type" `var`. An implicitly typed local variable is strongly typed as if you had declared the type yourself, but the compiler determines the type. The following two declarations of `i` are functionally equivalent:
```
Comment (IEvangelist) [inline docs/csharp/language-reference/statements/declarations.md:17]:
```suggestion
Beginning with C# 3, variables that are declared at method scope can have an implicit "type" `var`. An implicitly typed local variable is strongly typed as if you had declared the type yourself, but the compiler determines the type. The following two declarations of `a` and `b` are functionally equivalent:
```

File: docs/csharp/language-reference/statements/declarations.md [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+```csharp
+var i = 10; // Implicitly typed.
+int i = 10; // Explicitly typed.
```
Comment (IEvangelist) [inline docs/csharp/language-reference/statements/declarations.md:21]:
```suggestion
var a = 10; // Implicitly typed.
int b = 10; // Explicitly typed.
```

File: docs/csharp/language-reference/statements/declarations.md [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+Subsequent usage of `p` is the same as using the variable returned by `GetContactInformation` because `p` is an alias for that variable. Changes to `p` also change the variable returned from `GetContactInformation`.
+
+You can access a value by reference in the same way. In some cases, accessing a value by reference increases performance by avoiding a potentially expensive copy operation. For example, the following statement shows how one can define a ref local value that is used to reference a value.
+
+```csharp
+ref VeryLargeStruct reflocal = ref veryLargeStruct;
+```
+
+The `ref` keyword is used both before the local variable declaration *and* before the value in the second example. Failure to include both `ref` keywords in the variable declaration and assignment in both examples results in compiler error CS8172, "Can't initialize a by-reference variable with a value."
```
Comment (IEvangelist) [inline docs/csharp/language-reference/statements/declarations.md:76]: Could we link to compiler error CS8172 here?

File: docs/csharp/language-reference/statements/declarations.md [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+```csharp
+ref VeryLargeStruct reflocal = ref veryLargeStruct; // initialization
+refLocal = ref anotherVeryLargeStruct; // reassigned, refLocal refers to different storage.
+```
+
+ Ref local variables must still be initialized when they're declared.
```
Comment (IEvangelist) [inline docs/csharp/language-reference/statements/declarations.md:83]:
```suggestion
Ref local variables must still be initialized when they're declared.
```

**Review by IEvangelist — APPROVED** [review]
This looks great, well done. I had a few nits, questions, and suggestions.

**Comment by Youssef1313** [comment]
> @Youssef1313 Is there a tag for the F1 help for `scoped`? I'll add it to this PR.

I'm not sure. I added a test in https://github.com/dotnet/roslyn/pull/64094 and will see what CI says :)

File: docs/csharp/language-reference/statements/declarations.md (hunk 10 above)
Comment (BillWagner) [reply docs/csharp/language-reference/statements/declarations.md:76]: not yet. We don't have help on that one yet.

**Comment by BillWagner** [comment]
closing and reopening for a fresh build

**Pull request merged**